src/ripple/core/sql_backend/SQLInterface_sqlite.cpp

```diff
@@ -0,0 +1,1280 @@
+    }
+
+    /* else use shard database */
+    return do_single(
+        LEDGER, ledgerIndex, [&](soci::session& session, LedgerIndex index) {
+            return ripple::getHashesByIndex(
+                session, j, ledgerIndex, ledgerHash, parentHash);
+        });
+}
+
+std::map<LedgerIndex, std::pair<uint256, uint256>>
+SQLInterface_sqlite::getHashesByIndex(
+    beast::Journal const& j,
+    LedgerIndex minSeq,
+    LedgerIndex maxSeq)
+{
+    std::map<LedgerIndex, std::pair<uint256, uint256>> ret;
+
+    /* if database exists, use it */
+    if (exists(LEDGER))
+    {
+        auto db = checkoutDb(LEDGER);
+        ripple::getHashesByIndex(*db, j, minSeq, maxSeq, ret);
+        return ret;
+    }
+
+    /* else use shard databases */
+    while (minSeq <= maxSeq)
+    {
+        LedgerIndex shardMaxSeq = lastLedgerSeq(seqToShardIndex(minSeq));
+        if (shardMaxSeq > maxSeq)
+            shardMaxSeq = maxSeq;
+        do_single(
+            LEDGER, minSeq, [&](soci::session& session, LedgerIndex index) {
+                ripple::getHashesByIndex(session, j, minSeq, shardMaxSeq, ret);
+                return true;
+            });
+        minSeq = shardMaxSeq + 1;
+    }
+
+    return ret;
+}
+
+Json::Value
+SQLInterface_sqlite::loadTxHistory(Application& app, LedgerIndex startIndex)
+{
+    Json::Value txs;
+
+    /* if databases exists, use it */
+    if (exists(TRANSACTION))
+    {
+        auto db = checkoutDb(TRANSACTION);
+        ripple::loadTxHistory(*db, app, txs, startIndex, 20, false);
+        return txs;
+    }
+
+    /* else use shard databases */
+    int quantity = 20;
+    iterate_back(
+        TRANSACTION,
+        (LedgerIndex)-1,
+        [&](soci::session& session, LedgerIndex index) {
+            int total = ripple::loadTxHistory(
+                session, app, txs, startIndex, quantity, true);
+            if (total > 0)
+            {
+                quantity -= total;
+                if (quantity <= 0)
+                    return false;
+                startIndex = 0;
+            }
+            else
+            {
+                startIndex += total;
+            }
+            return true;
+        });
+
+    return txs;
+}
+
+NetworkOPs::AccountTxs
+SQLInterface_sqlite::getAccountTxs(
+    Application& app,
+    LedgerMaster& ledgerMaster,
+    beast::Journal& j,
+    AccountID const& account,
+    std::int32_t minLedger,
+    std::int32_t maxLedger,
+    bool descending,
+    std::uint32_t offset,
+    int limit,
+    bool bUnlimited)
+{
+    NetworkOPs::AccountTxs ret;
+
+    /* if databases exists, use it */
+    if (exists(TRANSACTION))
+    {
+        auto db = checkoutDb(TRANSACTION);
+        ripple::getAccountTxs(
+            *db,
+            app,
+            ledgerMaster,
+            j,
+            account,
+            ret,
+            minLedger,
+            maxLedger,
+            descending,
+            offset,
+            limit,
+            -1,
```

Comment:
`-1`? What's going on here?

Reply:
Replaced -1 by empty std::optional<int>. This exceptional value means that all data will be read from single DB so some simplification may took place.